src/mame/mit/cadr_tv_control.cpp

```diff
@@ -0,0 +1,161 @@
+// license:BSD-3-Clause
+// copyright-holders:Wilbert Pol
+/**********************************************************************************
+
+CADR TV Control emulation
+
+The TV control contains sync ram which looks like it can be configured to
+generate output to different types of TV displays.
+
+This currently emulates a 768x939 monochrome display.
+
+Other information about displays:
+
+Color screen: 454x576, 16 colors, RGB.
+Grayscale monitor: 454x576, 16 scales of gray.
+'CPT' monitor: 768x896, monochrome?
+
+60MHz and 64Mhz clocks are mentioned in the docs, as well as 64.69Hz default refresh rate.
+
+Implementation based on description of the operation, not schematics.
+
+
+**********************************************************************************/
+#include "emu.h"
+#include "cadr_tv_control.h"
+
+#include "screen.h"
+
+
+//#define VERBOSE (LOG_GENERAL)
+#include "logmacro.h"
+
+
+DEFINE_DEVICE_TYPE(CADR_TV_CONTROL, cadr_tv_control_device, "cadr_tv_control", "CADR TV Control")
+
+
+cadr_tv_control_device::cadr_tv_control_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, CADR_TV_CONTROL, tag, owner, clock)
+	, m_screen(*this, "screen")
+	, m_irq_cb(*this)
+{
+}
+
+void cadr_tv_control_device::device_add_mconfig(machine_config &config)
+{
+	SCREEN(config, m_screen, SCREEN_TYPE_RASTER);
+	m_screen->set_refresh_hz(60);
+	m_screen->set_vblank_time(0);
+	m_screen->set_screen_update(FUNC(cadr_tv_control_device::screen_update));
+	m_screen->set_size(SCREEN_WIDTH, SCREEN_HEIGHT);
+	m_screen->set_visarea_full();
+}
+
+
+uint32_t cadr_tv_control_device::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	const u32 black = 0x000000;
+	const u32 white = 0xffffff;
+
+	for (int y = 0; y < SCREEN_HEIGHT; y++)
+	{
+		const u16 line_start = y * (SCREEN_WIDTH / 32);
+
+		for (int x = 0; x < (SCREEN_WIDTH / 32); x++)
+		{
+			const u32 d = m_video_ram[line_start + x];
+			const u32 xs = x * 32;
+
+			for (int i = 0; i < 32; i++)
+			{
+				bitmap.pix(y, xs + i) = BIT(d, i) ? white : black;
+			}
+		}
+	}
```

Comment:
Please restrict the update to the clipping rectangle, at least in the vertical direction.